state/protocol/badger/state.go

```diff
@@ -259,6 +259,10 @@ func (state *State) bootstrapStatePointers(root protocol.Snapshot) func(*badger.
 		if err != nil {
 			return fmt.Errorf("could not insert sealed height: %w", err)
 		}
+		err = operation.IndexBySealedBlockID(segment.Seal.ID(), segment.Seal.BlockID)(tx)
```

Comment:
Actually, I think what I did is right, and could work for both root snapshot and non-root snapshot.

Note: for both cases, we will finalize and seal exactly one block, rather than multiple blocks. In other words, we only need to index for the lowest block height, rather than each seal in the payload of the highest block.

Let's walk though each of them. 
1. for root snapshot like this [R], we only finalize and seal one block, and the same block. 
2. for non-root snapshot, let's look at a happy path first: `A <- B (seal_X) <- C (seal_Y)<- D (seal_A)`, in this case, we set `D` as [the finalized height as well as the root height](https://github.com/onflow/flow-go/blob/b02e05c7e2e5c4c2bf8ee2a8de00d9128ae3c645/state/protocol/badger/state.go#L250-L254). And then we set `A` as the sealed height. 
Note: we didn't finalize A, B, C, but only D. and we didn't seal B, C, D yet. So we only finalize and seal one block.
Note: the sealed height is lower than the root height.  (side question, this should be OK, right? no one is supposed to query state for a height that is lower than root height)

3. let's look at another case where the last block has multiple seals: `A <- B (seal_X) <- C (seal_Y) <- D(seal_Z, seal_A)`. in this case, we do exactly the same as case `2`, finalize D and seal A, which exactly finalize and seal one block.
Note: this case is not possible: `A <- B (seal_X) <- C (seal_Y) <- (seal_A, seal_B)`. because this is an invalid sealing segment. The highest seal which is seal_B does not match with the lowest block which is A. 

Since we only seal one block, and is the lowest block, then it makes sense to index seal for the lowest block. And since one of the future blocks will have a seal for block B (the next unsealed block), and once that block is finalized, then we will index seal_B by block B. 

I also [listed the cases in comments for valid and invalid segments](https://github.com/onflow/flow-go/pull/2551/files#diff-04d7b3c448f8446615a26c7bc5ebe1ace7a6d4f4105cd5ea42ae8832cdc61767R267-R272), can go through that to validate my assert.

Reply:
Thanks Leo for the comment.
You are stating the the following 
A <- B (seal_X) <- C (seal_Y) <- (seal_A, seal_B)
would not be a valid sealing segment. To the best of my understanding, Supporting this case is necessary. But we can figure this out independently of this PR. 

Essentially, you are saying that the sealing segment can only seal one block. Ok. Then I think we should remove the extra 'Seal' field which you add in this PR. Because the sealing segment already specifies a single seal, which you argue is the only one. So no need to add another one.

Reply:
> Thanks Leo for the comment.
> You are stating the the following 
> A <- B (seal_X) <- C (seal_Y) <- (seal_A, seal_B)
> would not be a valid sealing segment. To the best of my understanding, Supporting this case is necessary. But we can figure this out independently of this PR. 
> 
> Essentially, you are saying that the sealing segment can only seal one block. Ok. Then I think we should remove the extra 'Seal' field which you add in this PR. Because the sealing segment already specifies a single seal, which you argue is the only one. So no need to add another one.

Reply:
A sealing segment for a reference block `D` is defined as a segment of blocks such that the latest sealed block w.r.t. `D` is the lowest block in the segment ([reference docs](https://github.com/onflow/flow-go/blob/3104f4a591072c9ed5a43fbd5767a1466ee89a93/model/flow/sealing_segment.go#L15-L18)).

```
A <- B (seal_X) <- C (seal_Y) <- D(seal_A, seal_B)
```
In this example, the latest sealed block w.r.t. `D` is `B`, not `A`, therefore:
* `A<-B<-C<-D` is not a valid sealing segment
* `B<-C<-D` is a valid sealing segment (the only valid sealing for reference block D)